pkg/scheduler/internal/queue/scheduling_queue_test.go

```diff
@@ -1499,13 +1500,21 @@ func TestPriorityQueue_MoveAllToActiveOrBackoffQueueWithQueueingHint(t *testing.
 			expectedQ: unschedulablePods,
 		},
 		{
-			name:    "QueueHintFunction is not called when Pod is gated",
-			podInfo: setQueuedPodInfoGated(&framework.QueuedPodInfo{PodInfo: mustNewPodInfo(p), UnschedulablePlugins: sets.New("foo")}),
+			name:    "QueueHintFunction is not called when Pod is gated by SchedulingGates plugin",
+			podInfo: setQueuedPodInfoGated(&framework.QueuedPodInfo{PodInfo: mustNewPodInfo(p), UnschedulablePlugins: sets.New(names.SchedulingGates, "foo")}),
 			hint: func(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) (framework.QueueingHint, error) {
 				return framework.Queue, fmt.Errorf("QueueingHintFn should not be called as pod is gated")
 			},
 			expectedQ: unschedulablePods,
 		},
+		{
+			name:    "QueueHintFunction is called when Pod is gated by a plugin other than SchedulingGates",
+			podInfo: setQueuedPodInfoGated(&framework.QueuedPodInfo{PodInfo: mustNewPodInfo(p), UnschedulablePlugins: sets.New("foo")}),
+			hint:    queueHintReturnQueue,
+			// FIXME: This should be backoffQ.
+			// https://github.com/kubernetes/kubernetes/issues/125538
+			expectedQ: activeQ,
```

Comment:
Found one minor bug, which can be follow-up.
https://github.com/kubernetes/kubernetes/issues/125538

Reply:
I guess this is intended, Pods rejected by preEnqueue shouldn't enter into backoffQ, because semantically the Pod hasn't been scheduled yet. cc @Huang-Wei

Reply:
No, it's definitely a bug. 
That behavior, assuming "Pods rejected by preEnqueue shouldn't enter into backoffQ, because semantically the Pod hasn't been scheduled yet", is only correct for the schedulinggate plugin, but not correct for any other PreEnqueue plugins.

Reply:
That's what I mean is for intended. https://github.com/kubernetes/enhancements/blob/master/keps/sig-scheduling/3521-pod-scheduling-readiness/README.md#implementation

```
A Pod can be moved to activeQ only when all Enqueue plugins returns Success. 
Otherwise, it's moved and parked in the internal unschedulable Pods pool
```

Would like to hear from Wei.

Reply:
Do we say that PreEnqueue plugins, **_in general_**, cannot change their decision once after they allow Pods to go to scheduling cycles? I couldn't find that specification.
So, what could happen, in my understanding, is:
1. A Pod that isn't gated goes to scheduling cycles, which means a Pod is supposed to go through backoff before proceeding to activeQ when it's retried.
2. Something gets changed and a custom PreEnqueue plugin now gates the Pod.
3. Something gets changed again, and a custom PreEnqueue plugin now allows the Pod to get scheduling again.
4. The Pod directly goes to activeQ, without going through backoff.

I know the scheduling gate plugin doesn't cause this because of-
> After creation, each schedulingGate can be removed in arbitrary order, but addition of a new scheduling gate is disallowed.
https://github.com/kubernetes/enhancements/blob/master/keps/sig-scheduling/3521-pod-scheduling-readiness/README.md#notesconstraintscaveats-optional

Reply:
I don't know if we have such case, but if a pod is gated again, it means it's not ready, maybe it should restart the scheduling lifecycle.. my two cents. Let's track this under your issue then.